### PR TITLE
Add dropout1d/2d/3d api

### PR DESCRIPTION
**Pull request by BBuf**
解决：https://github.com/Oneflow-Inc/libai/issues/342#issuecomment-1207565463 

- [x] 注册 nn.Dropout1d/2d/3d api。
- [x] 添加文档和接口测试。 

图片

图片

图片

### CODE REVIEW

File: python/oneflow/nn/__init__.py [hunk 1]
```diff
@@ -95,6 +95,9 @@
 )
 
 from oneflow.nn.modules.dropout import Dropout
+from oneflow.nn.modules.dropout import Dropout1d
```
Comment (Flowingsun007) [inline python/oneflow/nn/__init__.py:98]:
```suggestion
from oneflow.nn.modules.dropout import Dropout, Dropout1d, Dropout2d, Dropout3d
```

File: python/oneflow/framework/docstr/dropout.py [hunk 2]
```diff
@@ -97,3 +97,263 @@
  
     """,
 )
+
+add_docstr(
+    oneflow.nn.Dropout,
```
Comment (Flowingsun007) [inline python/oneflow/framework/docstr/dropout.py:102]: 常用class的docstr感觉直接写在class里就行（没必要单独拿出来写？）

File: python/oneflow/nn/modules/dropout.py [hunk 3]
```diff
@@ -131,6 +56,14 @@ def forward(self, x, addend=None):
             addend=addend if addend is not None else None,
         )
 
+class Dropout1d(Dropout):
+    pass
```
Comment (Flowingsun007) [inline python/oneflow/nn/modules/dropout.py:60]: 和torch对齐一下，顺手加一下F.dropoutXd接口的导出吧，然后这里forward时：
```python
def forward(self, input: Tensor) -> Tensor:
        return F.dropoutXd(input, self.p, self.training, self.inplace)
```

**Comment by BBuf** [comment]
对于PyTorch的Dropout1d/2d/3d来说，PyTorch均是使用了Bernoulli和广播乘以及Squeeze/Unsqueeze等Kernel进行拼接，没有写一个固定的fuse kernel，这个pr先对齐PyTorch的做法在Functor层实现这三个接口。

File: python/oneflow/nn/__init__.py (hunk 1 above)
Comment (BBuf) [reply python/oneflow/nn/__init__.py:98]: 好的

File: python/oneflow/framework/docstr/dropout.py (hunk 2 above)
Comment (BBuf) [reply python/oneflow/framework/docstr/dropout.py:102]: 都可以吧，卷积的话也是单独拿出来写的

File: python/oneflow/nn/modules/dropout.py (hunk 3 above)
Comment (BBuf) [reply python/oneflow/nn/modules/dropout.py]: 已经重构了dropout1d/2d/3d的实现，和pytorch实现一致。

File: oneflow/core/functional/impl/nn_functor.cpp [hunk 4]
```diff
@@ -14,9 +14,12 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+#include <cstdint>
```
Comment (MARD1NO) [inline oneflow/core/functional/impl/nn_functor.cpp:17]: 不必要的头文件都删了

File: oneflow/core/functional/functional_api.yaml [hunk 5]
```diff
@@ -808,7 +808,7 @@
   signature:
     [
       "Tensor (Tensor input, *, DataType dtype=kFloat, Generator generator=None) => Bernoulli",
-      "Tensor (Tensor input, Double p, *, DataType dtype=kFloat, Generator generator=None) => Bernoulli",
+      "Tensor (Tensor input, Double p, *, DataType dtype=kFloat, Generator generator=None) => BernoulliProb",
```
Comment (Flowingsun007) [inline oneflow/core/functional/functional_api.yaml:811]: 感觉Bernoulli没必要改名吧

File: oneflow/core/functional/functional_api.yaml [hunk 6]
```diff
@@ -1793,6 +1793,18 @@
   signature: "Tensor (Tensor dy, Tensor mask, Float scale) => DropoutGrad"
   bind_python: False
 
+- name: "dropout1d"
+  signature: "Tensor (Tensor input, Float p=0.5, Bool train=True) => Dropout1d"
```
Comment (Flowingsun007) [inline oneflow/core/functional/functional_api.yaml:1797]: 看了下，dropout1/2/3d的参数和torch没对齐：`train -> training`；另外少了`inplace` 
https://pytorch.org/docs/1.12/generated/torch.nn.functional.dropout2d.html#torch.nn.functional.dropout2d

File: oneflow/core/functional/functional_api.yaml (hunk 5 above)
Comment (BBuf) [reply oneflow/core/functional/functional_api.yaml:811]: 这里要调用这个functor，如果不改名则只能调用到 Tensor (Tensor input, *, DataType dtype=kFloat, Generator generator=None)

File: oneflow/core/functional/functional_api.yaml (hunk 6 above)
Comment (BBuf) [reply oneflow/core/functional/functional_api.yaml:1797]: 好的。training我对齐一下，这里暂时不能支持inplace因为里面涉及到squeeze/unsqueeze/emty/BernoulliProb都没有支持inplace，可以后续支持了再改。

File: oneflow/core/functional/functional_api.yaml (hunk 5 above)
Comment (Flowingsun007) [reply oneflow/core/functional/functional_api.yaml:811]: 好的

File: oneflow/core/functional/functional_api.yaml (hunk 6 above)
Comment (Flowingsun007) [reply oneflow/core/functional/functional_api.yaml]: 好的，可以标记个TODO

**Review by Flowingsun007 — APPROVED** [review]

**Review by marigoold — APPROVED** [review]
